specs/interop/fault_proof.md

```diff
@@ -9,16 +9,292 @@
 
 <!-- END doctoc generated TOC please keep comment here to allow auto update -->
 
-TODO
+## `SuperRoot`
+
+The `SuperRoot` is an aggregate proposal of L2 superchain state to L1,
+from which other L2 activity can be proven.
+
+The `SuperRoot` represents a snapshot of the superchain:
+it aggregates the output-roots of individual L2 chains, aligned to a single global timestamp.
+
+A `SuperSnapshot` is defined as the following SSZ data-structure:
+```python
+MAX_SUPERCHAIN_SIZE = 2**20 # the binary merkle-tree is 20 levels deep (excluding the SSZ length-mixin)
+
+class SuperSnapshot(Container:
+  chains: List[OutputRoot, MAX_SUPERCHAIN_SIZE]
+  timestamp: uint64
+```
+
+For each `OutputRoot`, the corresponding L2 block must be the last `safe` `block` 
+such that `block.timestamp <= snapshot.timestamp`.
+Each `OutputRoot` must be a [version 1 output root](#output-root-version-1).
+
+The output-roots must be ordered by ascending chain ID, with exactly one output-root
+per L2 chain in the Superchain.
+
+The `SuperRoot` is computed from SSZ hash-tree-root of the snapshot, versioned with a zero prefix byte:
+`0x00 ++ hash_tree_root(super_snapshot)[1:]`,
+where `||` is concatenation of bytes and `[1:]` slices out the first byte of the hash-tree-root.
+The `hash_tree_root` is computed with the `keccak256` hash function rather than `SHA2` as in the beacon-chain,
+for improved integration into the EVM.
+
+The fault-proof is contained to the L1 view of the `SuperRoot` claim
+(all L1 history up to and including the L1 timestamp when the claim was made),
+and does not include later-included batch data.
+
+## `OutputRoot` version 1
+
+We adapt the output-root that commits to the state of an individual L2 like in "version 0",
+but now extended with a `messages_root`.
+The `messages_root` commits to accumulators of the initiating messages and executing messages.
```

Comment:
I like this idea of `messages_root` - curious if you have thought about the structure at all yet. I think a sparse merkle tree is most simple where the key is the blockhash and the leaf is a vector commitment to the executing (ssz'ified subset of tx data) + initiating messages (all logs emitted)